Enforce valid field mapping exists for timestamp_field in templates.

martijnvg opened the pull request:
This change is just about validating that composable templates have a `date` or `date_nanos` field mapping configured in the mapping section for the timestamp_field that is defined in the data stream definition. 

A followup change will store the timestamp_field mapping with the data stream instance in the cluster state, so that the timestamp mapping will stick with the data stream as long as it will exist. This will make sure subsequent rollover invocations always create a new backing index with consistent timestamp field mapping.

Relates to #53100

martijnvg committed:
Enforce valid field mapping exists for timestamp_field in templates.

Relates to #53100

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Data streams)

martijnvg committed:
removed incorrect console snippet

martijnvg committed:
skip validating if no data stream template

martijnvg committed:
fixed docs snippets

martijnvg committed:
fixed unit tests

martijnvg committed:
fixed snippet

martijnvg committed:
incase of data stream template also generate a mapping

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -442,9 +444,20 @@ public ClusterState addIndexTemplateV2(final ClusterState currentState, final bo
+        ClusterState newState = ClusterState.builder(currentState)
             .metadata(Metadata.builder(currentState.metadata()).put(name, finalIndexTemplate))
             .build();
+        if (finalIndexTemplate.getDataStreamTemplate() != null) {
+            validateDataStreamTemplate(name, finalIndexTemplate, newState);
+        }
```
This validation needs to be moved to the `validateCompositeTemplate` method, that way, when a component template is updated we will still validate the mapping is correct.

Otherwise a composable template could be added where the mapping is in the component template, and then the component template could be updated to remove the timestamp mapping. Once we move the validation it'll be checked for both index and component templates

martijnvg replied:
Thanks for catching this. I've placed the validation in the wrong place.

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -442,9 +444,20 @@ public ClusterState addIndexTemplateV2(final ClusterState currentState, final bo
+
+    private void validateDataStreamTemplate(String name, ComposableIndexTemplate finalIndexTemplate, ClusterState state) throws Exception {
+        String tsFieldName = finalIndexTemplate.getDataStreamTemplate().getTimestampField();
+        Map<String, Object> finalMapping = parseV2Mappings("{}", resolveMappings(state, name), xContentRegistry);
```
When moving the validation to the `validateCompositeTemplate` method we should be able to reuse the mapping generated in that method

dakrone reviewed (changes requested):
Thanks for adding this Martijn, I left a comment, we need to move the validation so it happens every time any kind of template is added or updated

martijnvg committed:
Merge remote-tracking branch 'es/master' into enforce_ts_field_in_templates

martijnvg committed:
moved timestamp_field validation to the right place inside the template service

martijnvg committed:
fixed test

martijnvg committed:
fixed test

danhermann reviewed (approved):
LGTM

dakrone reviewed (changes requested):
Thanks for moving the validation @martijnvg, I think it's in the correct place now. I found another issue though, where we would not support nested timestamp fields, which I think we may want to support, what do you think?

martijnvg committed:
Merge remote-tracking branch 'es/master' into enforce_ts_field_in_templates

martijnvg commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:
```diff
@@ -158,4 +163,20 @@ public static ComposableIndexTemplate lookupTemplateForDataStream(String dataStr
+
+    public static void validateTimestampFieldMapping(String timestampFieldName, Map<?, ?> mapping) {
+        String timestampFieldMapperPath = "properties." + timestampFieldName;
+        Map<?, ?> timestampFieldMapper = ObjectPath.eval(timestampFieldMapperPath, mapping);
+        if (timestampFieldMapper == null) {
+            throw new IllegalArgumentException("expected timestamp field [" + timestampFieldName + "], but found no timestamp field");
+        }
```
Let me check, I assumed that `ObjectPath.eval(...)` handles nested fields.

martijnvg replied:
🤦

martijnvg committed:
move constants to top of file

martijnvg committed:
Use MapperService to validate the timestamp_field's field mapping instead of custom parsing logic.

This avoids introducing custom parsing logic, at the cost
of using mapper service in tests, which imo is minimal.

martijnvg committed:
fixed tests

martijnvg committed:
Merge remote-tracking branch 'es/master' into enforce_ts_field_in_templates

martijnvg committed:
fixed test

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java:
```diff
@@ -158,4 +165,16 @@ public static ComposableIndexTemplate lookupTemplateForDataStream(String dataStr
+        }
+        String type = timestampFieldMapper.typeName();
+        if (ALLOWED_TIMESTAMPFIELD_TYPES.contains(type) == false) {
+            throw new IllegalArgumentException("expected timestamp field [" + timestampFieldName + "] to be of types [" +
+                ALLOWED_TIMESTAMPFIELD_TYPES + "], but instead found type [" + type  + "]");
```
Super minor:

```suggestion
            throw new IllegalArgumentException("expected timestamp field [" + timestampFieldName + "] to be of types " +
                ALLOWED_TIMESTAMPFIELD_TYPES + ", but instead found type [" + type  + "]");
```

To avoid the `[[]]` double brackets

dakrone reviewed (approved):
LGTM, I left one minor comment, can we add a test case for the nested case `foo.timestamp` also?

martijnvg commented:
Thanks @dakrone for the review! The commit I added did include a unit test for the nested case: [testValidateNestedTimestampFieldMapping()](https://github.com/elastic/elasticsearch/pull/57741/files#diff-dd18f242501656ea717626ac6241c259R181) Is that sufficient? Or did you mean an integration test?

martijnvg committed:
adjusted error message

martijnvg committed:
Merge remote-tracking branch 'es/master' into enforce_ts_field_in_templates
